src/bitcoin/electrum/mod.rs

```diff
@@ -0,0 +1,259 @@
+/// Interface for Electrum backend.
+pub struct Electrum {
+    client: ElectrumClient,
+    bdk_wallet: wallet::BdkWallet,
+}
+
+impl Electrum {
+    pub fn new(
+        electrum_config: &config::ElectrumConfig,
+        network: &bitcoin::Network,
+        main_descriptor: &LianaDescriptor,
+    ) -> Result<Self, ElectrumError> {
+        let client = ElectrumClient::new(&electrum_config.addr.to_string(), network)?;
+        let genesis_hash = client.genesis_block().hash;
+        let bdk_wallet = wallet::BdkWallet::new(main_descriptor, genesis_hash);
+
+        Ok(Self { client, bdk_wallet })
+    }
+
+    pub fn client(&self) -> &ElectrumClient {
+        &self.client
+    }
+
+    pub fn wallet_coins(&self, outpoints: Option<&[OutPoint]>) -> HashMap<OutPoint, Coin> {
+        self.bdk_wallet.coins(outpoints)
+    }
+
+    pub fn init_wallet(
+        &mut self,
+        tip: &BlockChainTip,
+        coins: &[Coin],
+        txs: &[bitcoin::Transaction],
+        receive_index: ChildNumber,
+        change_index: ChildNumber,
+    ) {
+        self.bdk_wallet
+            .init(tip, coins, txs, receive_index, change_index)
+    }
+
+    pub fn sync_wallet(
+        &mut self,
+        receive_index: ChildNumber,
+        change_index: ChildNumber,
+    ) -> Result<(), ElectrumError> {
+        self.bdk_wallet
+            .sync_with_electrum(&self.client.0, receive_index, change_index)
+            .map_err(ElectrumError::Server)
+    }
+
+    pub fn common_ancestor(&self, tip: &BlockChainTip) -> Option<BlockChainTip> {
+        self.bdk_wallet
+            .common_ancestor_with_electrum(&self.client.0, tip)
+    }
+
+    pub fn wallet_transaction(
+        &self,
+        txid: &bitcoin::Txid,
+    ) -> Option<(bitcoin::Transaction, Option<Block>)> {
+        self.bdk_wallet.get_transaction(txid)
+    }
+}
+
+// If Electrum takes more than 3 minutes to answer one of our queries, fail.
+const RPC_SOCKET_TIMEOUT: u8 = 180;
+
+// Number of retries while communicating with the Electrum server.
+// A retry happens with exponential back-off (base 2) so this makes us give up after (1+2+4+8+16+32=) 63 seconds.
+const RETRY_LIMIT: u8 = 6;
+
+pub struct ElectrumClient(Client);
+
+impl ElectrumClient {
+    /// Create a new client and perform sanity checks.
+    pub fn new(url: &str, network: &bitcoin::Network) -> Result<Self, ElectrumError> {
+        let config = Config::builder()
+            .retry(RETRY_LIMIT)
+            .timeout(Some(RPC_SOCKET_TIMEOUT))
+            .build();
+        let client = bdk_electrum::electrum_client::Client::from_config(url, config)
+            .map_err(ElectrumError::Server)?;
+        let ele_client = Self(client);
+        ele_client.sanity_checks(network)?;
+        Ok(ele_client)
+    }
+
+    fn sanity_checks(&self, network: &bitcoin::Network) -> Result<(), ElectrumError> {
+        let server_features = self.0.server_features().map_err(ElectrumError::Server)?;
+        log::debug!("{:?}", server_features);
```

Comment:
Let's have a sentence for context if we want to keep this log post debugging.

Reply:
Ah yes, this was a leftover from testing really. I'll modify it.

Reply:
I've removed this and the call to `server_features` (as mentioned in https://github.com/wizardsardine/liana/pull/1222#discussion_r1711528045).